core/src/main/java/org/apache/iceberg/FilteredManifest.java

```diff
@@ -95,60 +94,38 @@ public FilteredManifest caseSensitive(boolean isCaseSensitive) {
       Evaluator evaluator = evaluator();
       InclusiveMetricsEvaluator metricsEvaluator = metricsEvaluator();
 
-      return CloseableIterable.filter(reader.entries(projection(fileSchema, columns, caseSensitive)),
+      // ensure stats columns are present for metrics evaluation
+      boolean requireStatsProjection = requireStatsProjection();
+      Collection<String> projectColumns = requireStatsProjection ? withStatsColumns(columns) : columns;
+
+      CloseableIterable<ManifestEntry> entries = CloseableIterable.filter(
+          reader.entries(projection(fileSchema, projectColumns, caseSensitive)),
           entry -> entry != null &&
               evaluator.eval(entry.file().partition()) &&
               metricsEvaluator.eval(entry.file()));
 
+      // note that columns itself could have stats projected, that's ok.
+      // We only drop stats if we have forced stats projection.
+      return CloseableIterable.transform(entries, e -> requireStatsProjection ? e.copyWithoutStats() : e);
```

Comment:
For performance, I think it is worse to copy here than to return the extra columns. We don't copy because we assume that the caller will use the row immediately -- e.g. to aggregate stats -- or will make a defensive copy. The caller is responsible to decide. That really cuts down on object allocation and helps performance. Copying just to get rid of stats columns forces a copy, when the caller may not need one.

Let's just return what was read here. In `iterator`, where we are copying anyway we can do the copy with or without stats.

Reply:
Do we need to think about the consequences of returning a different schema than what was projected?

Reply:
It's okay as long as we return more columns than the user asked for. There's an argument that you shouldn't return columns that weren't projected because of the extra memory required, but I think that's a minor concern.